File: ai server/aiserver.py
import requests
import json
from flask import Flask, request, jsonify
import os


# Initialize Flask App
app = Flask(__name__)

# Groq API Configuration
GROQ_API_URL = "https://api.groq.com/openai/v1/chat/completions"
GROQ_API_KEY = GROQ_API_KEY = os.getenv("GROQ_API_KEY") # Replace with your actual API key

# Function to summarize reviews using Groq API
def summarize_reviews(place_name, reviews):
    if not reviews:
        return f"No reviews found for {place_name}."

    system_prompt = (
        "You are an AI assistant that summarizes reviews for places. "
        "Given a list of user reviews, generate a concise and insightful summary "
        "highlighting common themes (positive and negative) in 50 sentences."
    )

    user_prompt = f"Reviews for {place_name}: {reviews}\nProvide a 50 lines summary:"

    try:
        response = requests.post(
            GROQ_API_URL,
            headers={"Authorization": f"Bearer {GROQ_API_KEY}", "Content-Type": "application/json"},
            json={
                "model": "mixtral-8x7b-32768",  # or "llama3-8b-8192"
                "messages": [
                    {"role": "system", "content": system_prompt},
                    {"role": "user", "content": user_prompt},
                ],
                "temperature": 0.7,
            },
        )
        
        response_data = response.json()
        
        if response.status_code != 200:
            return f"API Error: {response_data.get('error', 'Unknown error')}"

        summary = response_data["choices"][0]["message"]["content"]
        return summary

    except Exception as e:
        return f"Error generating summary: {str(e)}"

# Home Route to Avoid 404
@app.route("/", methods=["GET"])
def home():
    return jsonify({"message": "AI Server is running. Use POST /get_summary to get summaries."})

# Ignore favicon requests from browsers
@app.route("/favicon.ico")
def favicon():
    return "", 204

# Flask Route: Get summarized reviews from a file or JSON
@app.route('/get_summary', methods=['POST'])
def get_summary():
    try:
        data = request.get_json()
        if not data:
            return jsonify({"error": "Invalid JSON format"}), 400

        place_name = data.get("place")
        if not place_name:
            return jsonify({"error": "Missing place name"}), 400

        # Load the scraped reviews from a file (Modify this path if needed)
        try:
            with open("scraped_reviews.json", "r", encoding="utf-8") as file:
                scraped_data = json.load(file)
        except FileNotFoundError:
            return jsonify({"error": "Scraped reviews file not found"}), 400

        # Find reviews for the requested place
        reviews = next((entry["reviews"] for entry in scraped_data if entry["place"] == place_name), None)

        if not reviews:
            return jsonify({"error": f"No reviews found for {place_name}"}), 400

        summary = summarize_reviews(place_name, reviews)
        return jsonify({"place": place_name, "summary": summary})

    except Exception as e:
        return jsonify({"error": f"Server error: {str(e)}"}), 500

if __name__ == '__main__':
    app.run(debug=True, port=5000)

File: ai server2/src/aiserver.ts
import express, { Request, Response } from "express";
import dotenv from "dotenv";
import axios from "axios";
import fs from "fs";
import path from "path";

dotenv.config();

const app = express();
app.use(express.json());

const GROQ_API_URL = "https://api.groq.com/openai/v1/chat/completions";
const GROQ_API_KEY = process.env.GROQ_API_KEY as string;

if (!GROQ_API_KEY) {
    console.error("Error: GROQ_API_KEY is missing. Check your .env file.");
    process.exit(1);
}

// ✅ Function to Read Reviews from `scraped_reviews.json`
const getReviewsFromFile = (placeName: string): string[] => {
    try {
        const filePath = path.join(__dirname, "scraped_reviews.json");
        console.log(`Reading reviews from: ${filePath}`);

        if (!fs.existsSync(filePath)) {
            console.error("❌ Error: scraped_reviews.json not found.");
            return [];
        }

        const rawData = fs.readFileSync(filePath, "utf-8");
        const data = JSON.parse(rawData);

        console.log(`Available Places: ${data.map((entry: any) => entry.place)}`); // Debugging

        const placeEntry = data.find((entry: any) => entry.place === placeName);
        if (placeEntry) {
            console.log(`✅ Reviews found for ${placeName}`);
            return placeEntry.reviews;
        } else {
            console.warn(`⚠️ No reviews found for ${placeName}`);
            return [];
        }
    } catch (error) {
        console.error("❌ Error reading reviews file:", error);
        return [];
    }
};

// ✅ Function to Summarize Reviews
const summarizeReviews = async (placeName: string): Promise<string> => {
    const reviews = getReviewsFromFile(placeName);

    if (!reviews || reviews.length === 0) {
        return `No reviews found for ${placeName}.`;
    }

    const systemPrompt = "You are an AI assistant that summarizes reviews for places...";
    const userPrompt = `Reviews for ${placeName}: ${reviews.join(" ")}\nProvide a summary:`;

    try {
        const response = await axios.post<{ choices: { message: { content: string } }[] }>(
            GROQ_API_URL,
            {
                model: "mixtral-8x7b-32768",
                messages: [
                    { role: "system", content: systemPrompt },
                    { role: "user", content: userPrompt },
                ],
                temperature: 0.7,
            },
            {
                headers: {
                    Authorization: `Bearer ${GROQ_API_KEY}`,
                    "Content-Type": "application/json",
                },
            }
        );

        return response.data.choices[0]?.message?.content || "No summary available.";
    } catch (error: any) {
        console.error("Axios error:", error);
        return `Error generating summary: ${error.message}`;
    }
};

// ✅ API Route to Get Summary from File
app.post("/get_summary", async (req: any, res: any) => {
    try {
        const { place } = req.body;

        if (!place) {
            return res.status(400).json({ error: "Missing place name" });
        }

        const summary = await summarizeReviews(place);
        res.json({ place, summary });

    } catch (error: any) {
        res.status(500).json({ error: `Server error: ${error.message}` });
    }
});

// ✅ Start Server
const PORT = process.env.PORT || 5000;
app.listen(PORT, () => console.log(`Server running on port ${PORT}`));
